### PR TITLE
chore(storybook): upgrade storybook from 6.5 to 8.0

### PR DESCRIPTION
**Pull request by stephenliang**
This pull request upgrades Storybook from 6.5 which was last published on May 22, 2022 to the latest version of Storybook 8.0 likely to be published around [March 11, 2024](https://github.com/storybookjs/storybook/discussions/25276). This upgrade will resolve multiple dependabot issues including some criticals due to our webpack version being pinned to a specific minor (since 6.5 doesn't support past that version). Additionally, we will be completely up-to-date with our usage of Storybook and likely not need migrations for quite some time.

The upgrade involved a couple steps:

1. Execution of the Storybook codemod auto-upgrader
2. Replacement of the unsupported current usage of dynamic story creation (which violated Storybook's default export linter rule as well) with a supported (albeit experimental) [indexer](https://storybook.js.org/docs/api/main-config-indexers) solution.
3. When replacement was not possible due to aliasing, I refactored the story to adhere to Storybook's CSFv3 standards (many of our dynamic stories iterated args as stories, wheras we can use the `argTypes` prop to select the specific arg needed)
4. Readjusting our dependency tree due to our usage of transitives that Storybook previously provided.
   1. remark-parse 8.0.3
   2. unified 9.2.0
5. Upgrade of the RTL addon to remove transitive legacy storybook dependencies
6. Removal of the storybook addon-options dependency as this functionality was merged into storybook core

The above dependency pins will be removed when `redactable-markdown` is upgraded.

## Things to Note

1. In Storybook v6, docs were per story. However, in Storybook v7 aggregates all docs and puts it into a "docs" story
2. Dynamic stories are an anti-pattern in Storybook and is not typically supported. Although I add support for dynamic stories, we should avoid using it. Many stories can get by using the controls/args feature in Storybook instead. To denote dynamic stories, they will now be titled with a suffix `.dynamicStory` in the title. These stories should be deprecated.

### Why not use Storybook 7.x?

1. Storybook 7.6 conflicts with our Karma webpack configuration externally resulting in the following error. This bug previously occurred for other [Storybook versions ](https://www.github.com/storybookjs/storybook/issues/21316)and was fixed but seems to have come back up. I have [created a bug](https://www.github.com/storybookjs/storybook/issues/26325) with Storybook to address this issue and enable upgrading to the 7.6 branch. However, the storybook maintainers have mentioned that 8.0 fixes this issue.

Note that Storybook is typically compiled using its own webpack configuration via the `.storybook/main.js` storybook manager. Compilation of Storybook in this manner does not appear to be an officially supported method and is subject to breakage. See "Karma tests for Storybook" below for further discussion.

```
 {
    moduleIdentifier: '/drone/src/apps/node_modules/@storybook/addon-actions/dist/index.js',
    moduleName: './node_modules/@storybook/addon-actions/dist/index.js',
    loc: '8:286-293',
    message: 'Critical dependency: require function is used in a way in which dependencies cannot be statically extracted',
    moduleId: './node_modules/@storybook/addon-actions/dist/index.js',
    moduleTrace: [ [Object], [Object], [Object], [Object] ],
    details: '    at CommonJsRequireContextDependency.getWarnings (/drone/src/apps/node_modules/webpack/lib/dependencies/ContextDependency.js:102:18)\n' +
      '    at Compilation.reportDependencyErrorsAndWarnings (/drone/src/apps/node_modules/webpack/lib/Compilation.js:3132:24)\n' +
   }
```

### Karma tests for Storybook

Code.org currently tests Storybook by webpack compiling Storybook via Karma and lastly mounting the component to assert that the component mounts successfully. While this works for our limited use cases, it is not a supported use case and does miss out on errors that can occur on browsers. We should seek to migrate to using the official test runner which uses Playwright (browsers) and Jest to test our stories and remove our usage of Karma. Continuing to use this method of test execution is of limited value as it doesn't expose errors that are seen on client side rendering as evidenced in a [recent PR](https://github.com/code-dot-org/code-dot-org/pull/56602).

## Auto-Upgrade Script

The following prompts were selected during codemod auto-upgrade:

```
🔎 found a 'new-frameworks' migration:
╭ Automigration detected ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                                                                              │
│   We've detected your project is not fully setup with Storybook's 7 new framework format.                                                                                                    │
│                                                                                                                                                                                              │
│   Storybook 7 introduced the concept of frameworks, which abstracts configuration for renderers (e.g. React, Vue), builders (e.g. Webpack, Vite) and defaults to make integrations easier.   │
│                                                                                                                                                                                              │
│   Your project should be updated to use Storybook's framework: @storybook/react-webpack5. We can attempt to do this for you automatically.                                                   │
│                                                                                                                                                                                              │
│   Here are the steps this migration will do to migrate your project:                                                                                                                         │
│   - Remove the following dependencies:                                                                                                                                                       │
│         - * @storybook/builder-webpack5                                                                                                                                                      │
│   - * @storybook/manager-webpack5                                                                                                                                                            │
│   - Add the following dependencies:                                                                                                                                                          │
│         - * @storybook/react-webpack5                                                                                                                                                        │
│   - Remove the core.builder field in .storybook/main.js.                                                                                                                                     │
│                                                                                                                                                                                              │
│                                                                                                                                                                                              │
│   To learn more about the new framework format, see: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#new-framework-api                                                       │
│                                                                                                                                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

```
🔎 found a 'autodocsTrue' migration:
╭ Automigration detected ────────────────────────────────────────────────────────────────────────────╮
│                                                                                                    │
│   We've changed the configuration of autodocs (previous docsPage), so now the value:               │
│     - docs.autodocs: true -- means automatically create docs for every CSF file                    │
│     - docs.autodocs: 'tag' -- means only create autodocs for CSF files with the 'autodocs' tag     │
│     - docs.autodocs: false -- means never create autodocs                                          │
│                                                                                                    │
│   Based on your prior configuration,  we can set the `docs.autodocs` to keep your old behaviour:   │
│                                                                                                    │
│   docs: { autodocs: true }                                                                         │
│                                                                                                    │
│   More info: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#autodocs-changes      │
│                                                                                                    │
╰────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

```
╭ Migration check ran successfully ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                                               │
│   Successful migrations:                                                                                                                                      │
│                                                                                                                                                               │
│   new-frameworks, autodocsTrue                                                                                                                                │
│                                                                                                                                                               │
│   ─────────────────────────────────────────────────                                                                                                           │
│                                                                                                                                                               │
│   If you'd like to run the migrations again, you can do so by running 'npx storybook@next automigrate'                                                        │
│                                                                                                                                                               │
│   The automigrations try to migrate common patterns in your project, but might not contain everything needed to migrate to the latest version of Storybook.   │
│                                                                                                                                                               │
│   Please check the changelog and migration guide for manual migrations and more information: https://storybook.js.org/migration-guides/7.0                    │
│   And reach out on Discord if you need help: https://discord.gg/storybook                                                                                     │
│                                                                                                                                                               │
│   ─────────────────────────────────────────────────                                                                                                           │
│                                                                                                                                                               │
│   Critical: The following dependencies are duplicated and WILL cause unexpected behavior:                                                                     │
│                                                                                                                                                               │
│   @storybook/core-events:                                                                                                                                     │
│   5.3.21, 7.0.26, 7.6.17                                                                                                                                      │
│                                                                                                                                                               │
│                                                                                                                                                               │
│                                                                                                                                                               │
│                                                                                                                                                               │
│   Attention: The following dependencies are duplicated which might cause unexpected behavior:                                                                 │
│                                                                                                                                                               │
│   @storybook/channels:                                                                                                                                        │
│   5.3.21, 7.6.17                                                                                                                                              │
│                                                                                                                                                               │
│                                                                                                                                                               │
│                                                                                                                                                               │
│                                                                                                                                                               │
│   You can find more information for a given dependency by running yarn why <package-name>                                                                     │
│                                                                                                                                                               │
│                                                                                                                                                               │
│                                                                                                                                                               │
│                                                                                                                                                               │
│   Please try de-duplicating these dependencies by running yarn dedupe                                                                                         │
│                                                                                                                                                               │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

```

The above mentioned duplicate dependencies were addressed as mentioned in the previous section by upgrading the parent dependencies.

## Links

- [jira ticket](https://codedotorg.atlassian.net/browse/P20-748)
- [storybook 6.5 to 8.0 migration guide](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)
- [Dynamic story creation reference](https://github.com/storybookjs/storybook/issues/9828#issuecomment-1953231029)
- [Storybook indexers documentation](https://storybook.js.org/docs/api/main-config-indexers)

## Testing story

I spun up the current [6.5 Storybook ](https://code-dot-org.github.io/cdo-styleguide/) and did a manual spot check of each story. I verified dynamic stories as well as static stories to be working.

## PR Checklist:

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [x] New features are translatable or updates will not break translations
- [x] Relevant documentation has been added or updated
- [x] User impact is well-understood and desirable
- [x] Pull Request is labeled appropriately
- [x] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/.storybook/main.js [hunk 1]
```diff
@@ -1,28 +1,67 @@
...
+const staticDirs = envConstants.STORYBOOK_STATIC_ASSETS
+  ? ['../build/package/', '../../dashboard/public']
+  : [];
 
-module.exports = {
-  stories: ['../src/**/*.story.@(js|jsx|ts|tsx)'],
+export default {
+  stories: [
+    '../src/**/*.story.@(js|jsx|ts|tsx)',
+    '../src/**/*.dynamicStory.@(js|jsx|ts|tsx)',
```
Comment (stephenliang) [inline apps/.storybook/main.js:14]: Stories that are to be dynamically generated should use the special `dynamicStory` file name to indicate its usage in the dynamic story indexer. Note, I would recommend we avoid the usage of dynamic stories going forward as it is not the[ best practice for creating stories in Storybook](https:/www./github.com/storybookjs/storybook/issues/9828#issuecomment-1370291568) (the method we use will be removed in v8).

File: apps/.storybook/preview.js [hunk 2]
```diff
@@ -1,19 +1,13 @@
...
   options: {
     storySort: {
-      order: [
-        'DesignSystem',
-        'templates',
-        'code-studio',
-      ],
-    }
-  }
+      method: 'alphabetical',
```
Comment (stephenliang) [inline apps/.storybook/preview.js:9]: This line replaces the `'@storybook/addon-options'` addon (the functionality was merged to core)

File: apps/src/componentLibrary/button/Button.excludedStory.tsx [hunk 3]
```diff
@@ -1,4 +1,3 @@
-/* eslint-disable storybook/default-exports */
```
Comment (stephenliang) [inline apps/src/componentLibrary/button/Button.excludedStory.tsx:1]: This story was renamed as Storybook complained that this file is missing default exports due to it being a completely commented file. I renamed it to no longer match the Storybook story glob to maintain the code.

**Review by levadadenys — COMMENTED** [review]
Super happy to see this update, thank you very much!
LGTM in general!
The only thing I'm concerned about is auto generated docs being not fully functional/generated. (Meaning for some reason it's not fetching the JSdocs props and component description into the documentation page)

**Comment by stephenliang** [comment]
I pushed an update to use the slower but more thorough typescript doc generator config! This should fix the missing documentation on Storybook. I also just updated to use the official version of Storybook 8.0, which was just released ~40 minutes ago!

**Review by levadadenys — APPROVED** [review]
Docs issue is now fixed.
LGTM! Thanks again!

File: apps/.storybook/main.js (hunk 1 above)
Comment (bencodeorg) [reply apps/.storybook/main.js:14]: Can you clarify the situation where you couldn't refactor (I think "When replacement was not possible due to aliasing,...") from your description?

File: apps/package.json [hunk 4]
```diff
@@ -297,6 +301,7 @@
     "rehype-raw": "^4.0.0",
     "rehype-react": "^3.1.0",
     "rehype-sanitize": "^2.0.2",
+    "remark-parse": "~8.0.3",
```
Comment (bencodeorg) [inline apps/package.json:304]: For my understanding, can you help me understand how you picked out these "transient dependency" versions (is that the right word)?

File: apps/script/deploy-storybook.sh [hunk 5]
```diff
@@ -29,7 +29,7 @@ popd
 
 # build the static storybook site
 echo "Building the static storybook site"
-NODE_OPTIONS="--max-old-space-size=4096" npx build-storybook -o $DIR_TO_DEPLOY
+NODE_OPTIONS="--max-old-space-size=4096" npx storybook build -o $DIR_TO_DEPLOY
```
Comment (bencodeorg) [inline apps/script/deploy-storybook.sh:32]: Were you able to run this script without the deploy to github steps?

File: apps/.storybook/unplugin.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+      const config = await serverRequire(fileName);
+      return await compile(config, options);
+    },
+  };
+});
+
+export const {esbuild} = unplugin;
+export const {webpack} = unplugin;
+export const {rollup} = unplugin;
```
Comment (bencodeorg) [inline apps/.storybook/unplugin.js:30]: Do you need all of these exports?

**Review by bencodeorg — COMMENTED** [review]
This is awesome! Might be a new record for being the closest to the latest release of a dependency :). A couple of questions just to make sure I'm following: 

1. re: karma/Storybook tests, am I right that we'll lose a bit of test coverage for the new `dynamicStory` entries?
2. I had a separate comment about dynamic stories, but just wanted to makes sure I fully understand the situation...am I reading right that there only 3 cases currently? Would it be hard to refactor? Seems like a fair amount of complexity we're adding to support a pattern that we don't want to perpetuate?

**Comment by bencodeorg** [comment]
(hah sorry for the post-merge comments, was writing them up as you merged!)

File: apps/package.json (hunk 4 above)
Comment (stephenliang) [reply apps/package.json:304]: Yes, for `remark-parse`:

Resolved version: [In this yarn.lock](https://github.com/code-dot-org/code-dot-org/blob/f0e9c36e0bcd7a14090458ef34c6002a94980bce/apps/yarn.lock#L25304)

### Before Storybook upgrade:

```
yarn why remark-parse
├─ @code-dot-org/redactable-markdown@npm:0.4.0
│  └─ remark-parse@npm:6.0.3 (via npm:^6.0.0)
│
├─ @mdx-js/mdx@npm:1.6.22
│  └─ remark-parse@npm:8.0.3 (via npm:8.0.3)
│
└─ remark-mdx@npm:1.6.22
   └─ remark-parse@npm:8.0.3 (via npm:8.0.3)
```

```
yarn why remark-mdx
└─ @mdx-js/mdx@npm:1.6.22
   └─ remark-mdx@npm:1.6.22 (via npm:1.6.22)
```

```
yarn why @mdx-js/mdx
└─ @storybook/mdx1-csf@npm:0.0.1
   └─ @mdx-js/mdx@npm:1.6.22 (via npm:^1.6.22)
```

```
yarn why @storybook/mdx1-csf
├─ @storybook/addon-docs@npm:6.5.17-alpha.0
│  └─ @storybook/mdx1-csf@npm:0.0.1 (via npm:^0.0.1)
│
├─ @storybook/addon-docs@npm:6.5.17-alpha.0 [4a0fb]
│  └─ @storybook/mdx1-csf@npm:0.0.1 (via npm:^0.0.1)
│
├─ @storybook/csf-tools@npm:6.5.17-alpha.0
│  └─ @storybook/mdx1-csf@npm:0.0.1 (via npm:^0.0.1)
│
└─ @storybook/csf-tools@npm:6.5.17-alpha.0 [2e816]
   └─ @storybook/mdx1-csf@npm:0.0.1 (via npm:^0.0.1)
```

---

### After upgrade:


```
yarn why remark-parse

└─ @code-dot-org/redactable-markdown@npm:0.4.0
   └─ remark-parse@npm:6.0.3 (via npm:^6.0.0)
```


Net result: Storybook removed the dependency on remark-parse 8.0 but Code.org uses remark-parse 8.0 features. Therefore, I needed to add a direct dependency on remark-parse 8.0.

File: apps/script/deploy-storybook.sh (hunk 5 above)
Comment (stephenliang) [reply apps/script/deploy-storybook.sh:32]: Yes, and was able to verify the Github page is also updated!

File: apps/.storybook/unplugin.js (hunk 6 above)
Comment (stephenliang) [reply apps/.storybook/unplugin.js:30]: No - as we don't use those other build systems. However, as a plugin it does make sense to export these from a plugin perspective.

**Comment by stephenliang** [comment]
> * re: karma/Storybook tests, am I right that we'll lose a bit of test coverage for the new `dynamicStory` entries?

Yes this is correct the dynamic tests won't be mounted because Karma isn't able to see them anymore (we were previously able to because of storyStoreV6, but we don't have that anymore). That being said, I'm fairly certain the Karma tests for storybook do not provide a lot of value. It only [mounts the component](https://github.com/code-dot-org/code-dot-org/blob/f0e9c36e0bcd7a14090458ef34c6002a94980bce/apps/test/storybook/util/testStorybook.js#L16) but doesn't assert anything. Even if the mount fails, it won't error out (this is evidenced by #56602 breaking Storybook but we didn't find out via the Karma tests.

Perhaps we should investigate using [Storybook's official test runner](https://storybook.js.org/docs/writing-tests/test-runner/) feature which will give us the ability to more deeply inspect the stories and create browser rich test interactions.

> 2. I had a separate comment about dynamic stories, but just wanted to makes sure I fully understand the situation...am I reading right that there only 3 cases currently? Would it be hard to refactor? Seems like a fair amount of complexity we're adding to support a pattern that we don't want to perpetuate?

Yes, here is the stack trace. I did refactor the tests that cause the below stack trace to no longer use dynamicStory and instead conform to the Storybook standard story format, there were 3 remaining that I enabled usage of dynamic stories but I didn't want to migrate them over to being a standard story given I haven't reached out to those folks. However I think a follow up could be to refactor them to no longer being dynamic stories and having us remove the dynamic plugin. Let me know what you think and I can kick off a cleanup effort.

```
Error: Cannot find module '@cdo/locale'
Require stack:
- /home/sliang/git-workspaces/code-dot-org/apps/src/templates/progress/BubbleFactory.jsx
- /home/sliang/git-workspaces/code-dot-org/apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx
- /home/sliang/git-workspaces/code-dot-org/apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.dynamicStory.jsx
- /home/sliang/git-workspaces/code-dot-org/apps/node_modules/@storybook/core-common/dist/index.js
- /home/sliang/git-workspaces/code-dot-org/apps/node_modules/@storybook/telemetry/dist/index.js
- /home/sliang/git-workspaces/code-dot-org/apps/node_modules/@storybook/cli/dist/generate.js
- /home/sliang/git-workspaces/code-dot-org/apps/node_modules/@storybook/cli/bin/index.js
- /home/sliang/git-workspaces/code-dot-org/apps/node_modules/storybook/index.js
```

File: apps/package.json (hunk 4 above)
Comment (bencodeorg) [reply apps/package.json:304]: Cool, TIL about `yarn why`! Two more questions for you for my learning:
1. How did you identify we use remark-parse 8.0 features? Test failures?
2. How did you identify which packages we might need to add direct dependencies for? Just anything where a major version dependency changed in `yarn.lock` after doing this?

**Comment by snickell** [comment]
Followup note: based on the size of the delta it looks like `yarn.lock` might have been regenerated from scratch here? No worries either way, but see: https://codedotorg.slack.com/archives/C0T0PNTM3/p1710870867280539

**Pull request merged**